library/ziti.c

```diff
@@ -992,83 +1063,119 @@ static void update_identity_data(ziti_identity_data *data, const ziti_error *err
-        uv_timer_stop(&ztx->refresh_timer);
-    }
+static void session_post_auth_query_cb(ziti_context ztx, int status, void *ctx) {
+    ZTX_LOG(VERBOSE, "post auth query callback starting with status[%s]", ziti_errorstr(status));
+    if (status == ZITI_OK) {
+        ZTX_LOG(VERBOSE, "transitioning to fully authenticated");
+        ziti_api_session *session = ztx->api_session;
 
-    ziti_posture_init(ztx, 20);
+        ziti_set_fully_authenticated(ztx);
+
+        update_ctrl_status(ztx, ZITI_OK, NULL);
+
+        ziti_ctrl_current_identity(&ztx->controller, update_identity_data, ztx);
+
+        //if we had auth queries, refresh state to zero out
+        if (session->auth_queries != NULL && *session->auth_queries != NULL) {
+            ziti_ctrl_current_api_session(&ztx->controller, update_session_data, ztx);
+        }
 
-    if (!ztx->no_current_edge_routers) {
-        ziti_ctrl_current_edge_routers(&ztx->controller, edge_routers_cb, ztx);
+        int time_diff;
+        if (session->cached_last_activity_at) {
+            ZTX_LOG(TRACE, "API supports cached_last_activity_at");
+            time_diff = (int) (ztx->session_received_at.tv_sec - session->cached_last_activity_at->tv_sec);
+        } else {
+            ZTX_LOG(TRACE, "API doesn't support cached_last_activity_at - using updated");
+            time_diff = (int) (ztx->session_received_at.tv_sec - session->updated->tv_sec);
+        }
+        if (abs(time_diff) > 10) {
+            ZTX_LOG(ERROR, "local clock is %d seconds %s UTC (as reported by controller)", abs(time_diff),
+                    time_diff > 0 ? "ahead" : "behind");
+        }
+
+        if (session->expires) {
+            // adjust expiration to local time if needed
+            session->expires->tv_sec += time_diff;
+            ZTX_LOG(DEBUG, "ziti api session expires in %ld seconds", (long) (session->expires->tv_sec - ztx->session_received_at.tv_sec));
+            long delay = (session->expires->tv_sec - ztx->session_received_at.tv_sec) - 10;
+            uv_timer_start(&ztx->session_timer, session_refresh, delay * 1000, 0);
+        }
+
+        if (ztx->opts->refresh_interval > 0 && !uv_is_active((const uv_handle_t *) &ztx->refresh_timer)) {
+            ZTX_LOG(DEBUG, "refresh_interval set to %ld seconds", ztx->opts->refresh_interval);
+            services_refresh(&ztx->refresh_timer);
+        } else if (ztx->opts->refresh_interval == 0) {
+            ZTX_LOG(DEBUG, "refresh_interval not specified");
+            uv_timer_stop(&ztx->refresh_timer);
+        }
+
+        ziti_posture_init(ztx, 20);
+
+        if (!ztx->no_current_edge_routers) {
+            ziti_ctrl_current_edge_routers(&ztx->controller, edge_routers_cb, ztx);
+        }
+    } else {
+        ZTX_LOG(VERBOSE, "transitioning to unauthenticated, unhandled status[%s]", ziti_errorstr(status));
+        ziti_set_unauthenticated(ztx); //disable?
+        update_ctrl_status(ztx, ZITI_NOT_AUTHORIZED, NULL);
```

Comment:
I added a few more of these in related branches. I don't know if this matters or not.